ext/hook/uhook.c

```diff
@@ -103,32 +122,50 @@ static void dd_uhook_call_hook(zend_execute_data *execute_data, zend_object *clo
 }
 
 static bool dd_uhook_begin(zend_ulong invocation, zend_execute_data *execute_data, void *auxiliary, void *dynamic) {
-    (void) invocation;
-
     dd_uhook_def *def = auxiliary;
     dd_uhook_dynamic *dyn = dynamic;
 
-    dyn->hook_data = zend_objects_new(ddtrace_hook_data_ce);
-    object_properties_init(dyn->hook_data, ddtrace_hook_data_ce);
+    if (def->closure && def->closure != ZEND_CLOSURE_OBJECT(EX(func))) {
+        dyn->hook_data = NULL;
+        return true;
+    }
 
-    ZVAL_LONG(ddtrace_hookdata_property_id(dyn->hook_data), def->id);
-    ZVAL_ARR(ddtrace_hookdata_property_args(dyn->hook_data), dd_uhook_collect_args(execute_data));
+    dyn->hook_data = (dd_hook_data *)dd_hook_data_create(ddtrace_hook_data_ce);
+
+    dyn->hook_data->invocation = invocation;
+    ZVAL_LONG(&dyn->hook_data->property_id, def->id);
+    ZVAL_ARR(&dyn->hook_data->property_args, dd_uhook_collect_args(execute_data));
 
     if (def->begin && !def->running) {
+        dyn->hook_data->execute_data = execute_data;
+
         def->running = true;
         dd_uhook_call_hook(execute_data, def->begin, dyn->hook_data);
         def->running = false;
     }
+    dyn->hook_data->execute_data = NULL;
 
     return true;
 }
 
 static void dd_uhook_end(zend_ulong invocation, zend_execute_data *execute_data, zval *retval, void *auxiliary, void *dynamic) {
-    (void) invocation;
-
     dd_uhook_def *def = auxiliary;
     dd_uhook_dynamic *dyn = dynamic;
 
+    if (!dyn->hook_data) {
+        return;
+    }
+
+    ddtrace_span_data *span = dyn->hook_data->span;
+    if (span && span->duration != DDTRACE_DROPPED_SPAN && span->duration != DDTRACE_SILENTLY_DROPPED_SPAN) {
+        zval *exception_zv = ddtrace_spandata_property_exception(span);
+        if (EG(exception) && Z_TYPE_P(exception_zv) <= IS_FALSE) {
+            ZVAL_OBJ_COPY(exception_zv, EG(exception));
+        }
+
+        dd_trace_stop_span_time(span);
+    }
```

Comment:
Apparently, it was only called before if we dropped `.end`, and now we want to always call it.